components/uniset/src/bin/gen_properties.rs

```diff
@@ -0,0 +1,458 @@
+        UnicodeSet::from_inversion_list(inv_list)
+    } else {
+        let code_point_str = range_str;
+        let code_point: u32 = u32::from_str_radix(code_point_str, 16).unwrap();
+        let inv_list: Vec<u32> = vec![code_point, code_point + 1];
+        UnicodeSet::from_inversion_list(inv_list)
+    };
+
+    let props_vals = get_data_line_prop_vals(&line_parts);
+
+    let range = if let Ok(range) = range_result {
+        range
+    } else {
+        let inv_list: Vec<u32> = Vec::default();
+        UnicodeSet::from_inversion_list(inv_list).unwrap()
+    };
+    (range, props_vals)
+}
+
+fn get_code_point_prop_vals(
+    code_point: u32,
+    code_point_overrides: &HashMap<UnicodeSet, HashMap<String, String>>,
+    blocks: &HashMap<UnicodeSet, HashMap<String, String>>,
+    defaults: &HashMap<String, String>,
+) -> HashMap<String, String> {
+    // create the map of applicable property values, and initialize it to
+    // defaults for the whole Unicode range
+    let mut prop_vals: HashMap<String, String> = HashMap::new();
+    prop_vals.clone_from(defaults);
+
+    // determine if this code point matches any of the block-wide defaults
+    // ("overrides")
+    for (range, block_prop_vals) in blocks {
+        if range.contains(char::from_u32(code_point).unwrap()) {
+            let block_prop_vals_clone = block_prop_vals.iter().map(|(k, v)| (k.clone(), v.clone()));
+            prop_vals.extend(block_prop_vals_clone);
+        }
+    }
+
+    // finally, apply the overrides for this code point
+    for (range, code_point_prop_vals) in code_point_overrides {
+        if range.contains(char::from_u32(code_point).unwrap()) {
+            let code_point_override_prop_vals_clone = code_point_prop_vals
+                .iter()
+                .map(|(k, v)| (k.clone(), v.clone()));
+            prop_vals.extend(code_point_override_prop_vals_clone);
+        }
+    }
+
+    prop_vals
+}
+
+fn get_binary_prop_unisets(
+    prop_aliases: &HashMap<String, HashSet<String>>,
+    code_points: &HashMap<u32, HashMap<String, String>>,
+) -> HashMap<String, UnicodeSet> {
+    let mut m: HashMap<String, UnicodeSet> = HashMap::new();
+
+    for (canonical_name, all_names) in prop_aliases {
+        let mut uniset_builder = UnicodeSetBuilder::new();
+        for (code_point, code_point_prop_vals) in code_points {
+            let code_point_prop_names: HashSet<String> =
+                code_point_prop_vals.keys().cloned().collect();
+            if !all_names.is_disjoint(&code_point_prop_names) {
+                uniset_builder.add_char(char::from_u32(*code_point).unwrap());
+            }
+        }
+        if !&uniset_builder.is_empty() {
+            let uniset = uniset_builder.build();
+            m.insert(canonical_name.clone(), uniset);
+        }
+    }
+
+    m
+}
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let filename = &args[1];
+
+    if let Ok(lines) = read_lines(filename) {
```

Comment:
Make `main()` return `Result<(), Error>` (for some `Error`, which could be local to this file), and then you can use `?`.  That's much nicer than manually handling errors via nested if statements.

https://www.joshmcguigan.com/blog/custom-exit-status-codes-rust/

Reply:
Done.